### PR TITLE
Provide asynchronous processing via NATS Streaming.

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis <alexellis2@gmail.com>

## Description

Provide asynchronous invocation and deferred execution via NATS Streaming.

* Asynchronous invocations are possible through a new route, change `function/<name>` to `async-function/<name>`. This means any function can be invoked asynchronously. 

* To see the results of a function check `docker service logs func_queue-worker`

* Async processing means you get a 202 after work is queued, you need to use a wrapper function or call another function once your code has completed so that the result can be retrieved or stored

* NATS Streaming implementation is in a new repository @ https://github.com/alexellis/faas-nats

* ./deploy_extended.sh and docker-compose.extended.yml contain the additional Swarm services to run a stack in asynchronous mode.

## Motivation and Context

Allows long-running tasks and background jobs.

This shows the difference between client interaction with sync/async function invocations.

screen shot 2017-08-18 at 5 57 51 pm

Pros:
* The result is returned to the caller immediately

Cons:
* The work could time-out

screen shot 2017-08-18 at 5 57 34 pm

Pros:
* Very fast response when ingesting work onto queue
* The work will be completed at some time in the future
* Any work can be scheduled this way

Cons:
* Harder to get result - you need to make sure your function stores or communicates its result with a backend service or another helper function.

## How Has This Been Tested?

Tested by user in asyc_nats branch for machine-learning. This is a manual merge of those changes to master and has been tested locally on Docker Swarm. 

The following gist (replace the branch with update_async) contains instructions on testing.

https://gist.github.com/alexellis/62dad83b11890962ba49042afe258bb1

Provides feature in issue: https://github.com/alexellis/faas/issues/103

- [x] Tested on Docker Swarm
- [x] Tested on Kubernetes
- [ ] Configuration for ARM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: gateway/Dockerfile [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+
+FROM alpine:latest
 WORKDIR /root/
 
 EXPOSE 8080
 ENV http_proxy      ""
 ENV https_proxy     ""
 
-COPY gateway    .
+COPY --from=0 /go/src/github.com/alexellis/faas/gateway/gateway    .
```
Comment (johnmccabe) [inline gateway/Dockerfile]: You can name the builder step for clarity:
```
FROM golang:1.8.3-alpine as BUILDER
COPY --from=builder ...
```

Comment (alexellis) [inline gateway/Dockerfile]: Sure.. Why not?

**Comment by rgee0** [comment]
> Async processing means you get a 200 after work is queued

I wonder whether 202 would be more semantically appropriate since the request has been accepted but not necessarily been actioned?

From [ietf rfc7231](https://tools.ietf.org/html/rfc7231#section-6.3.3)

> The 202 (Accepted) status code indicates that the request has been
   accepted for processing, but the processing has not been completed.
   The request might or might not eventually be acted upon, as it might
   be disallowed when processing actually takes place.  There is no
   facility in HTTP for re-sending a status code from an asynchronous
   operation.

 > The 202 response is intentionally noncommittal.  Its purpose is to
   allow a server to accept a request for some other process (perhaps a
   batch-oriented process that is only run once per day) without
   requiring that the user agent's connection to the server persist
   until the process is completed.  The representation sent with this
   response ought to describe the request's current status and point to
   (or embed) a status monitor that can provide the user with an
   estimate of when the request will be fulfilled.

**Comment by alexellis** [comment]
Let's do it

**Comment by alexellis** [comment]
How about testing end to end on your setup or digital Ocean?

**Comment by johnmccabe** [comment]
Seeing some odd distribution across 3 queue_workers, I may have not have built the queue-worker correctly (there was no `../gateway` for it to copy from), will have a look again tomorrow.
```
func_queue-worker.3.qwzphiwjpef3@ip-172-31-2-198.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.1.399841zoumlm@ip-172-31-2-198.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.3.qwzphiwjpef3@ip-172-31-2-198.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.2.v6fr1t66sx81@ip-172-31-3-76.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.2.v6fr1t66sx81@ip-172-31-3-76.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.2.v6fr1t66sx81@ip-172-31-3-76.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.2.v6fr1t66sx81@ip-172-31-3-76.eu-west-2.compute.internal    | Request for func_echoit.
func_queue-worker.2.v6fr1t66sx81@ip-172-31-3-76.eu-west-2.compute.internal    | Request for func_echoit.
<continues to hit worker 2>
```

**Comment by alexellis** [comment]
re: the `../gateway` part you can comment that out. I need to re-push the queue-worker build script.

**Comment by alexellis** [comment]
* Pending: Header X-Callback-Url passed along with request

Results in queue-worker invoking URL with result.

```
$ curl localhost:8080/async-function/func_hubstats -d "library" -H "X-Callback-Url: https://requestb.in/qg3fd4qg"
```

screen shot 2017-08-21 at 08 59 41


This additional work is being done in the queue-worker repository.

https://github.com/alexellis/nats-queue-worker

**Comment by muka** [comment]
Can we help / do any test on `nats-queue-worker` ?

**Comment by alexellis** [comment]
Yes of course @muka - end to end testing is perfect on Swarm and / or Kubernetes. It would be good to have a couple of functions which work together to demonstrate this new flow.

**Comment by muka** [comment]
Just a note, the gist describing how to test is not aligned, the `docker-compose.yml` ref in this branch point to an image called `:async-test` for gateway. `build.sh` creates latest-dev instead

**Comment by alexellis** [comment]
Hi there, the gist will now be slightly out of date but should help with getting the idea of how to test. I've updated the images in the YAML so you should not need to build from source now. Please could you give this a run-through? Have I invited you to Slack yet?

**Comment by Tofull** [comment]
@muka Here is the routine I used to test this awesome pull request : 
```bash
## Clone faas and go to PR#131
cd
git clone https://github.com/alexellis/faas
cd faas
git checkout update_async


## Add recent work on queue worker image from open-faas/nats-queue-worker
cd
git clone https://github.com/open-faas/nats-queue-worker
cd nats-queue-worker
git checkout tags/0.1
cd ..
mv nats-queue-worker/queue-worker faas/
rm -rf nats-queue-worker
cd faas


## Build queue-worker docker image
(cd queue-worker && ./build.sh)


## Build faas
./build.sh


## Deploy faas
docker stack rm func
docker swarm init #if needed
./deploy_extended.sh


## Test faas
curl localhost:8080/function/func_echoit/ -d "Should directly return this text"
# > ok

curl -v localhost:8080/async-function/func_echoit/ -d "You should see nothing expected a HTTP/1.1 202 Accepted"
# > ok

# be sure your async request has been well computed.
docker service logs -f func_queue-worker
# > ok (you should see the message "You should see nothing expected a HTTP/1.1 202 Accepted")

## Create a requestbin
# Go to https://requestb.in to create a requestbin url

## test your requestb with basic curl and refresh page to see how it works
# with POST
curl -X POST -d "hello=post" https://requestb.in/zk9eqqzk
# > ok

# with GET
curl -X GET -d "hello=get" https://requestb.in/zk9eqqzk
# > ok

# with GET and querystring
curl -X GET https://requestb.in/zk9eqqzk?hello=querystring
# > querystring

# with GET and no params
curl -X GET https://requestb.in/zk9eqqzk
# > ok (None)


## Try to call requestb with async faas
curl localhost:8080/async-function/func_echoit/ -d "test requestbin without params" -H "X-Callback-Url: https://requestb.in/zk9eqqzk"
# > Posting report - 200 but request bin not called

curl localhost:8080/async-function/func_echoit/ -d "test requestbin with params" -H "X-Callback-Url: https://requestb.in/zk9eqqzk?faas=isworking"
# > Posting report - 200 but request bin not called

docker service logs -f func_queue-worker


# or if you want to use another features like sms broadcaster (from Free provider)
curl localhost:8080/async-function/func_echoit/ -d "send sms when finished" -H "X-Callback-Url: https://smsapi.free-mobile.fr/sendmsg?user=USER_ID&pass=USER_PASS&msg=Finish%20Task%20!"
```
The last curl should send me a SMS when the task is done : 
async_task_send_me_sms

Make IoT with FaaS :tada:

**Comment by johnmccabe** [comment]
Thats awsome @Tofull !!

**Comment by alexellis** [comment]
I have tested this on Kubernetes over at Digital Ocean, it also works there now 👍

**Comment by alexellis** [comment]
Initial guide / documentation. @Tofull what else is useful?

https://github.com/alexellis/faas/blob/master/guide/asynchronous.md

**Comment by Tofull** [comment]
Maybe a link to this discussion : 
`As of [PR #131](https://github.com/alexellis/faas/pull/131) asynchronous invocation is available for testing.`

LGTM otherwise

**Comment by jmkhael** [comment]
**Proposal for: "a happy and a sad path callback url, based on a selector in json field"**

Basically, this would enable me to chain calls to a different URL if it "fails".

* X-Callback-Url-happy-path: url to post to if it succeeds
* X-Callback-Url-sad-path: url to post to if it fails

the definition of success/failure needs either to be via the status_codes, or via a field, in the json payload/result.

an example:
```
curl -d "http://jmkhael.ioi/" http://localhost:8080/async-function/url_ping -H "X-Callback-Url-happy-path: http://gateway:8080/function/slack_it -H "X-Callback-Url-sad-path: http://gateway:8080/function/sms
```

happy output - success (json.status === "ok")
```
{"url":"http://jmkhael.io/", "status": "ok", "status_code": "200"}
```

sad output - failure (json.status !== "ok")
```
{"url":"http://jmkhael.ioi/","status":"nok","message":"failed to ping http://jmkhael.ioi/","error":"HTTPConnectionPool(host='jmkhael.ioi', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb69dec49d0>: Failed to establish a new connection: [Errno -2] Name does not resolve',))"}
```

**Pull request merged**